### ISSUE TITLE
Add weighted_sum and weighted_sum_binary rules to model_organisms framework

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/cruijff_kit#460]
## Motivation

The model_organisms framework (`src/tools/model_organisms/`) currently supports rules whose output is a single token from a small discrete space (`parity`, `majority`, `first`, `last`, `nth`, `length`, `constant`, `coin`, `min`, `max`). It does not support rules where the output is an arithmetic function of the input.

`weighted_sum` and `weighted_sum_binary` add the linear-DGP shape, letting downstream researchers ask questions like "Does fine-tuning recover a known linear data-generating process? How does the answer depend on sparsity, noise level, model size, or sequence length?"

This issue is for **library/infra work**: adding the rules and supporting framework changes. Specific experiments using these rules are out of scope.

## New rules

Two rules in `src/tools/model_organisms/rules.py`, both `applicable={"bits", "digits"}`:

- **`weighted_sum`** — returns the integer `w·x + intercept`, formatted as a spaced signed zero-padded string.
- **`weighted_sum_binary`** — returns `"0"` or `"1"`. Deterministic threshold when `noise_scale == 0`; Bernoulli sample with `p = σ((w·x + intercept) / noise_scale)` when `noise_scale > 0`. Stochastic samples are deterministic per `(weight_seed, sequence)` — note this differs from `coin`, which keys on the dataset `seed`. Keying label noise on the DGP seed is deliberate (see "Seed semantics" below).

Using these rules with `letters` raises a clear error.

Both rules are also restricted to `design: memorization` and `design: in_distribution`. Using them with `design: ood` raises a clear error — see *Out of scope* below for the rationale.

## `rule_kwargs` schema

| Field | Type | Default | Description |
|---|---|---|---|
| `weights` | `list[int] \| None` | `None` | Explicit weight vector of length `k`. If given, distribution params are ignored. |
| `weight_max` | `int` | `3` | Weights drawn uniformly from `{-W, …, -1, 1, …, W}` (excluding zero). |
| `sparsity` | `float` | `0.0` | Bernoulli probability that a drawn weight is then masked to zero. Applied after the magnitude draw. |
| `weight_seed` | `int \| None` | `None` (uses dataset seed) | Seed for the **DGP**: weight draws, sparsity mask, and (for the binary rule) the per-sequence label-noise stream. See "Seed semantics" below. |
| `intercept` | `int \| "balanced"` | `0` | DGP intercept. `"balanced"` resolves to `round(-sum(w_i) · E[x_i])` over the input alphabet (4.5 for digits, 0.5 for bits). |
| `noise_scale` | `float` | `0.0` | Binary rule only. `0` = deterministic threshold; `> 0` = Bernoulli sample. |

## Seed semantics

These rules expose **two seeds** with separate roles, so that DGP definition can be decoupled from sequence sampling:

- **`seed`** (the existing positional argument to `generate()`): controls *which sequences you observe* — primary sequence sampling, OOD sequence sampling (when applicable), the dataset-seed-keyed `coin` rule, and the bayes_accuracy MC sampler. Same role as for every other rule.
- **`weight_seed`** (in `rule_kwargs`, defaults to `seed` when `None`): controls *what the DGP is and how its noise behaves* — magnitude draws, the sparsity mask, and (for `weighted_sum_binary` with `noise_scale > 0`) the per-sequence Bernoulli stream.

Useful patterns this enables:
- **Same DGP, fresh sequences** — fix `weight_seed`, vary `seed`. Resolved weights, intercept, format width, and (where present) the per-sequence label-noise pattern stay constant; the train/val sample changes. Useful for studying sample-size effects or for averaging FT outcomes over sampling noise.
- **Same sample, different DGP** — fix `seed`, vary `weight_seed`. Sequence sampling stays fixed; the DGP changes. Useful for varying weight sparsity / magnitude while controlling for which sequences land in train vs. val.

Note the deliberate difference from `coin`: `coin` keys label draws on `(seed, sequence)`, so changing `seed` reshuffles labels. `weighted_sum_binary` keys label draws on `(weight_seed, sequence)`, so changing `seed` does **not** reshuffle labels — only `weight_seed` does. This is what makes "same DGP, fresh sequences" a clean comparison.

## Output format

Spaced + signed + zero-padded integer string. Format width is computed once per dataset from the resolved weights and intercept: `digits = ceil(log10(sum(|w_i|) · alphabet_max + |intercept| + 1))`, plus a sign token.

Example: `k=8`, weights `[1, 2, -1, 3, 0, 1, -2, 1]`, digits alphabet → max output 99 → 2-digit width → output 42 renders as `+ 4 2`.

This generalizes across all major BPE/SentencePiece tokenizers (Llama-2/3, GPT-4, Claude, Mistral, Qwen) because whitespace is a hard tokenization boundary.

## Framework change: `Rule.prepare` hook

The current rule signature `fn(seq, **kwargs) -> str` is per-example. `weighted_sum` needs dataset-level state (resolved weights, format width, Bayes accuracy). Extend the `Rule` dataclass:

```python
@dataclass(frozen=True)
class Rule:
    name: str
    fn: Callable[..., str]
    applicable: frozenset[str]
    prepare: Callable[..., dict] | None = None
```

When non-None, `prepare(input_type, k, N, seed, rule_kwargs) -> dict` is called once at generation time and its return value is merged into `rule_kwargs` before per-example `fn` calls. `generate.py` is updated to invoke this hook. All existing rules keep `prepare=None` — no behavioral change.

## Dataset metadata

Top-level dataset metadata is extended for these rules to record:

- `resolved_weights: list[int]`
- `intercept: int`
- `noise_scale: float` (binary only)
- `format_width: int`
- `bayes_accuracy: float` (binary stochastic only) — analytical or Monte Carlo estimate over the input distribution

Downstream users get ground-truth weights for coefficient-recovery analyses and the optimal-model upper bound without recomputation.

## Out of scope (file as separate follow-up issues)

- ICL baseline infra (an `n_shot` parameter on `inspect_task.py`). Strongly recommended for any downstream experiment using these rules — without it, FT improvements are uninterpretable (Vacareanu 2024).
- Tokenizer sanity check at data generation. Generic feature, not specific to these rules.
- Continuous (non-integer) weights and float output formatting.
- Additional weight distributions beyond uniform-integer.
- **OOD design with `weighted_sum` / `weighted_sum_binary`.** The resolved weight vector, intercept, and format width are dataset-level state shared across all splits, which creates several footguns under `design: ood`:
  - per-split DGP overrides (e.g. different `weights` in `ood_tests[i].rule_kwargs`) would be silently ignored, since primary's `resolved_weights` propagate to every split;
  - OOD splits with `k ≠ k_primary` would surface as a per-row length-mismatch exception during dataset construction rather than as an upfront validation error;
  - OOD splits that change `input_type` (bits ↔ digits) would silently render outputs at a different width than primary, because the `+0Nd` formatter pads but doesn't truncate.

  Supporting OOD properly requires explicit design — re-drawing weights creates a different DGP (so it isn't an OOD test of the *same* DGP), padding with zeros makes an arbitrary extrapolation choice, and per-split format widths break the "stable output format across the dataset" guarantee. Until a concrete length-generalization or distribution-shift use case appears, `generate()` rejects `design: ood` for these rules at validation time. File a follow-up issue if a use case emerges.

## Acceptance criteria

- [ ] `weighted_sum` and `weighted_sum_binary` registered; returned by `list_rules()`
- [ ] `Rule.prepare` hook implemented and invoked from `generate.py`; existing rules unchanged
- [ ] Output format stable across all examples in a dataset
- [ ] Dataset metadata records resolved weights, intercept, noise_scale, format width, and Bayes accuracy (binary stochastic)
- [ ] `generate()` rejects `design: ood` for `weighted_sum` and `weighted_sum_binary` with a clear error message
- [ ] Unit tests cover: explicit weights, distribution-drawn weights (deterministic per seed), `sparsity=1.0` → all-zero weights, deterministic threshold, stochastic sample (deterministic per `(seed, sequence)`), `intercept="balanced"` produces approximately balanced binary classes, `applicable` rejects `letters`, `design: ood` rejected for both rules, edge cases (`k=1`, all-zero weights, `sparsity=0`), recorded `bayes_accuracy` matches empirical optimal-classifier accuracy within statistical tolerance
- [ ] Documentation updated: rule docstrings (note `majority` and `constant` as special cases), parameter table in `.claude/skills/design-experiment/references/model_organisms.md`, examples in `blueprints/model_organism/README.md`
- [ ] `ruff check` and `ruff format --check` pass on changed files

## References

- Vacareanu et al. (COLM 2024). "From Words to Numbers: Your LLM Is Secretly A Capable Regressor When Given In-Context Examples." [arxiv 2404.07544](https://arxiv.org/abs/2404.07544) — establishes that pretrained LLMs already do regression in-context, motivating the ICL-baseline follow-up.
- Singh & Strouse (2024). "Tokenization counts: the impact of tokenization on arithmetic in frontier LLMs." [arxiv 2402.14903](https://arxiv.org/abs/2402.14903) — anchors the spaced-format choice.
